Migrate algorithms to empty constructors and use ATH_MSG_XXX

kratsg opened the pull request:
See #635 for details as well as #887, #617.

# Changes

- `m_registered` now keeps track of whether a given instance was registered in the registry instance or not [only to prevent accidental multiple calls of `registerInstance()` on grid, but will not prevent this from happening between client and job]
- constructors are all empty, they do not take any parameters/arguments. `xAH::Algorithm` takes in a `className` which just defines the name of the class, for tracking in the instance registry
- (most) functions in `HelperFunctions` now want a `MsgStream` object passed in from the algorithm in order to print out information/messaging
- `m_debug` and `m_verbose` are no more
- verbosity of an algorithm is set using the `m_debugLevel` configuration which takes case-insensitive string values that are part of the `MSG::Level` enum: 
  - `nil` (0)
  - `verbose` (1)
  - `debug` (2) 
  - `info` (3)
  - `warning` (4)
  - `error` (5) 
  - `fatal` (6)
  - `always` (7)

  Most people will only use `verbose`, `debug`, and `warning`. `info` is used by default.
- to set an algorithm verbosity outside of `xAH_run.py`, you want to do something like

  ```python
  import ROOT; ROOT.gROOT.Macro("$ROOTCOREDIR/scripts/load_packages.C")
  ROOT.asg.ToolStore # fix because of https://its.cern.ch/jira/browse/ATLASG-270
  alg = ROOT.JetSelector()
  alg.msg().setLevel(ROOT.MSG.VERBOSE)
  ```

## To Do
- [x] migrate constructors to empty constructors and rely on `m_name` and `TNamed::SetName`
- [x] migrate some of the functions in `HelperFunctions` (the big ones such as `retrieve`)
- [x] update runner scripts to allow for `TNamed::SetName` calls automagically if `m_name` is a configurable option
- [x] require `m_name` as an option to any algorithm
- [x] finish up the rest of the functions in `HelperFunctions`
- ~[ ] get rid of any other Error/Info/Warning/Debug calls scattered throughout the code~
- [x] deprecate `m_debug` and `m_verbose` everywhere (direct references to `m_debug/verbose` shouldn't really be used for anything
- [x] update xAH_run/xAH_config to allow the verbosity of an algorithm to be configurable via `m_debugLevel`

johnda102 commented:
This will break all user code.

Why remove m_debug from the base class ?


On Wed, May 10, 2017 at 5:35 AM, Giordon Stark <notifications@github.com>
wrote:

> Assigned #882 <https://github.com/UCATLAS/xAODAnaHelpers/pull/882> to
> @johnda102 <https://github.com/johnda102>.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/UCATLAS/xAODAnaHelpers/pull/882#event-1076081454>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AGx2Cpu3yotO5Q1bESlGIZ9wlCzcEH5Sks5r4YTVgaJpZM4NUJEA>
> .
>



-- 
University of Chicago
http://hep.uchicago.edu/~johnda

kratsg commented:
> This will break all user code.

It's actually backwards-compatible for 90% of the changes.

> Why remove m_debug from the base class ?

Because it's been replaced by `this->msg().msgLevel(MSG::DEBUG)` which means we can set the level using `m_debugLevel`

johnda102 commented:
On Wed, May 10, 2017 at 10:03 AM, Giordon Stark <notifications@github.com>
wrote:

> This will break all user code.
>
> It's actually backwards-compatible for 90% of the changes.
>
> Why remove m_debug from the base class ?
>
> Because it's been replaced by this->msg().msgLevel(MSG::DEBUG) which
> means we can set the level using m_debugLevel
>
Why not have m_debug in the base xAH::Algorithm ?

This can then set
  this->msg().msgLevel(MSG::DEBUG)
on initialize if true ?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/UCATLAS/xAODAnaHelpers/pull/882#issuecomment-300491654>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AGx2Cvu1i9evscCXnLInYLcb5YABsqJVks5r4cOVgaJpZM4NUJEA>
> .
>



-- 
University of Chicago
http://hep.uchicago.edu/~johnda

johnda102 commented:
The 10% changes which are not backwards-compatible will breaks all user
code.

On Wed, May 10, 2017 at 10:05 AM, John Alison <johnda102@gmail.com> wrote:

>
>
> On Wed, May 10, 2017 at 10:03 AM, Giordon Stark <notifications@github.com>
> wrote:
>
>> This will break all user code.
>>
>> It's actually backwards-compatible for 90% of the changes.
>>
>> Why remove m_debug from the base class ?
>>
>> Because it's been replaced by this->msg().msgLevel(MSG::DEBUG) which
>> means we can set the level using m_debugLevel
>>
> Why not have m_debug in the base xAH::Algorithm ?
>
> This can then set
>   this->msg().msgLevel(MSG::DEBUG)
> on initialize if true ?
>>
>> —
>> You are receiving this because you were mentioned.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/UCATLAS/xAODAnaHelpers/pull/882#issuecomment-300491654>,
>> or mute the thread
>> <https://github.com/notifications/unsubscribe-auth/AGx2Cvu1i9evscCXnLInYLcb5YABsqJVks5r4cOVgaJpZM4NUJEA>
>> .
>>
>
>
>
> --
> University of Chicago
> http://hep.uchicago.edu/~johnda
>



-- 
University of Chicago
http://hep.uchicago.edu/~johnda

kratsg commented:
> Why not have m_debug in the base xAH::Algorithm ?

Because `m_debug` is currently used in the non-algorithm code (containers, histogrammers) and now has a different meaning than `m_debugLevel`.



> The 10% changes which are not backwards-compatible will breaks all user
> code.

You should try checking out this branch first, compiling it, and running it without any changes from you. It should work for pretty much all configurations. `m_debug` won't do anything, and `m_name` is required from a configurational standpoint.

If anyone is inheriting from `xAH::Algorithm`, their algorithms will still work.

If anyone is using xAH with the helper functions in their own algorithms, it will still work - but the logging will mention "HelperFunctions" instead of the algorithms they called it from.

And all the histogramming code has been primarily untouched, as has the container code. Info/Error/Warning/Debug are functions that still exist, but these are not in our algorithms, they can still be in other users' algorithms, and they can still use `m_debug` there.

kratsg commented:
I should also point out that these changes are the best compromise users are going to get because there will be an abrupt change due to the AnalysisBase release header changes (algorithm changes, etc...) and what I've done here can absorb it without breaking.

johnda102 commented:
On Wed, May 10, 2017 at 10:11 AM, Giordon Stark <notifications@github.com>
wrote:

> Why not have m_debug in the base xAH::Algorithm ?
>
> Because m_debug is currently used in the non-algorithm code (containers,
> histogrammers) and now has a different meaning than m_debugLevel.
>

I dont see the difference.

> The 10% changes which are not backwards-compatible will breaks all user
> code.
>
> You should try checking out this branch first, compiling it, and running
> it without any changes from you. It should work for pretty much all
> configurations. m_debug won't do anything, and m_name is required from a
> configurational standpoint.
>

I did and it failed to build, b/c of user code that inherits from
xAH::Algorithm

> If anyone is inheriting from xAH::Algorithm, their algorithms will still
> work.
>

They do not compile

> If anyone is using xAH with the helper functions in their own algorithms,
> it will still work - but the logging will mention "HelperFunctions" instead
> of the algorithms they called it from.
>

The interface also changed for retireve so these fail to compile

> And all the histogramming code has been primarily untouched, as has the
> container code. Info/Error/Warning/Debug are functions that still exist,
> but these are not in our algorithms, they can still be in other users'
> algorithms, and they can still use m_debug there.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/UCATLAS/xAODAnaHelpers/pull/882#issuecomment-300494168>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AGx2CoqPmfx_87gp_BpKqpKCfaXq9L3yks5r4cV6gaJpZM4NUJEA>
> .
>



-- 
University of Chicago
http://hep.uchicago.edu/~johnda

kratsg commented:
> I dont see the difference.

m_debug is a bool, m_debugLevel is an integer.

> I did and it failed to build, b/c of user code that inherits from
> xAH::Algorithm

xAH::Algorithm's signature hasn't changed. You should be able to inherit from it primarily. My guess is the missing `m_debug` broke it, so I added it back in for now, but it's being deprecated.

> The interface also changed for retrieve so these fail to compile

I'll make this backwards compatible.
